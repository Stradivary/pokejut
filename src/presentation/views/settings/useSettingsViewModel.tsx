import { useSimulator } from "@/domain/use-cases/simulator";
import { useMantineColorScheme } from "@mantine/core";
import { modals } from "@mantine/modals";
import { notifications } from "@mantine/notifications";
import { useCallback, useMemo } from "react";

export function useSettingsViewModel() {
  
  const colorSchemeOptions = [
    {
      value: "light",
      label: "Terang",
    },
    {
      value: "dark",
      label: "Gelap",
    },
  ];

  const { setColorScheme, colorScheme } = useMantineColorScheme();


  const { pokemonList, clearPokemonList, clearSelectedPokemon } =
    useSimulator();

  const canReleaseCollection = useMemo(
    () => pokemonList.length === 0,
    [pokemonList]
  );


  const handleReleaseCollection = useCallback(
    () =>
      onReleaseCollection(
        clearSelectedPokemon,
        clearPokemonList,
        notifications
      ),
    [clearPokemonList, clearSelectedPokemon]
  );


  return {
    colorScheme,
    setColorScheme,
    colorSchemeOptions,
    handleReleaseCollection,
    canReleaseCollection,
  };
}

export function onClearCache(setRerender, navigate) {
  localStorage.clear();
  notifications.show({
    title: "Berhasil",
    message: "Cache berhasil dihapus",
    color: "blue",
    icon: <img src="/pokeball.png" alt="pokeball" />,
  });
  setRerender((e) => e + 1);
  navigate("/settings", { replace: true });
}

export function onReleaseCollection(
  clearSelectedPokemon: () => void,
  clearPokemonList: () => void,
  notifications: any
) {
  modals.openConfirmModal({
    title: "Lepas Semua Pokemon",
    children: "Apakah kamu yakin ingin melepaskan semua koleksi pokemon?",
    onConfirm: () =>
      confirmReleaseCollection(
        clearSelectedPokemon,
        clearPokemonList,
        notifications
      ),
    labels: {
      confirm: "Lepas Semua",
      cancel: "Batal",
    },
  });
}

export function confirmReleaseCollection(
  clearSelectedPokemon: () => void,
  clearPokemonList: () => void,
  notifications: any
): void {
  clearSelectedPokemon();
  clearPokemonList();
  notifications.show({
    title: "Berhasil",
    message: "Koleksi Pokemon berhasil dihapus",
    color: "blue",
    icon: <img src="/pokeball.png" alt="pokeball" />,
  });
}
