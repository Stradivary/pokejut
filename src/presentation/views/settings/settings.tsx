import { Button, Divider, Group, SegmentedControl, Stack, Title } from "@mantine/core";
import { useSettingsViewModel } from "./useSettingsViewModel";

export function Component() {
    const bindings = useSettingsViewModel();

    return <Stack p={16}>
        <Title order={1} fz={18} >Pengaturan</Title>
        <Group w="80%" mt={16} justify="space-between">
            <Title order={2} fz={16} fw="bold">Tema</Title>
            <SegmentedControl
                value={bindings.colorScheme}
                onChange={bindings.handleChangeColorScheme}
                data={bindings.colorSchemeOptions} fullWidth />
        </Group>
        <Divider />

        <Title order={2} fz={16} fw="bold">Koleksi Pokemon</Title>

        <Button
            maw={240}
            disabled={bindings.canReleaseCollection}
            color="red"
            onClick={bindings.handleReleaseCollection}>
            Hapus Koleksi Pokemon
        </Button>

    </Stack>;
}

Component.displayName = 'SettingsPage';