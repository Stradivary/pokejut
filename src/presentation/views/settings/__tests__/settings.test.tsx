import { beforeAll, describe, expect, it, vi } from "vitest";
import snapshotSanitizer from "~/tests/snapshot-sanitizer";
import { render } from "~/tests/test-utils";
import { Component as Settings } from "../settings";
import {
  confirmReleaseCollection,
  onReleaseCollection
} from "../useSettingsViewModel";

vi.mock("../useSettingsViewModel", async () => {
  const actual = await vi.importActual("../useSettingsViewModel");
  return {
    ...actual,
    colorScheme: "light",
    setColorScheme: vi.fn(),
    colorSchemeOptions: ["light", "dark"],
    canReleaseCollection: false,
    handleReleaseCollection: vi.fn(),
  };
});

describe("Settings Component", () => {
  beforeAll(() => {
    expect.addSnapshotSerializer(snapshotSanitizer);
  });
  it("should render the component correctly", () => {
    const { container } = render(<Settings />);
    expect(container).toMatchSnapshot();
  });
});

describe("onReleaseCollection", () => {
  it("should call clearSelectedPokemon and clearPokemonList functions", () => {
    const clearSelectedPokemon = vi.fn();
    const clearPokemonList = vi.fn();
    const notifications = {
      show: vi.fn(),
    };

    onReleaseCollection(clearSelectedPokemon, clearPokemonList, notifications);

    expect(clearSelectedPokemon).toBeDefined();
  });

  it("should call confirmReleaseCollection function", () => {
    const clearSelectedPokemon = vi.fn();
    const clearPokemonList = vi.fn();
    const notifications = {
      show: vi.fn(),
    };

    confirmReleaseCollection(
      clearSelectedPokemon,
      clearPokemonList,
      notifications
    );

    expect(notifications).toBeDefined();
  });
});


