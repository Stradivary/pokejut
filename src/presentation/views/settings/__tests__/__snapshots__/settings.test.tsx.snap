// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings Component > should render the component correctly 1`] = `
<div>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, sans-serif;--mantine-font-family-headings: Poppins, sans-serif;--mantine-radius-default: calc(0.5rem * var(--mantine-scale));}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_6d731127 mantine-Stack-root"
    style="--stack-gap: var(--mantine-spacing-md); --stack-align: stretch; --stack-justify: flex-start;"
  >
    <h1
      class="m_8a5d1357 mantine-Title-root"
      data-order="1"
      style="--title-fw: var(--mantine-h1-font-weight); --title-lh: var(--mantine-h1-line-height); --title-fz: var(--mantine-h1-font-size);"
    >
      Pengaturan
    </h1>
    <div
      class="m_4081bf90 mantine-Group-root"
      style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: space-between; --group-wrap: wrap; width: 80%;"
    >
      <h2
        class="m_8a5d1357 mantine-Title-root"
        data-order="2"
        style="--title-fw: var(--mantine-h2-font-weight); --title-lh: var(--mantine-h2-line-height); --title-fz: var(--mantine-h2-font-size); font-weight: bold;"
      >
        Tema
      </h2>
      <div
        class="m_cf365364 mantine-SegmentedControl-root"
        data-full-width="true"
        data-initialized="true"
        data-with-items-borders="true"
        role="radiogroup"
        style="--sc-shadow: var(--mantine-shadow-xs);"
      >
        <span
          class="m_9e182ccd mantine-SegmentedControl-indicator m_96b553a6 mantine-FloatingIndicator-root"
          style="--transition-duration: var(--sc-transition-duration); transform: translateY(0px) translateX(0px); width: 0px; height: 0px;"
        />
        <div
          class="m_69686b9b mantine-SegmentedControl-control"
          data-active="true"
          style=""
        >
          <input
            class="m_1714d588 mantine-SegmentedControl-input"
            data-focus-ring="auto"
            id=""
            name=""
            style=""
            type="radio"
            value="light"
          />
          <label
            class="m_1738fcb2 mantine-SegmentedControl-label"
            data-active="true"
            for=""
            style=""
          >
            <span
              class="m_78882f40 mantine-SegmentedControl-innerLabel"
              style=""
            >
              Terang
            </span>
          </label>
        </div>
        <div
          class="m_69686b9b mantine-SegmentedControl-control"
          style=""
        >
          <input
            class="m_1714d588 mantine-SegmentedControl-input"
            data-focus-ring="auto"
            id=""
            name=""
            style=""
            type="radio"
            value="dark"
          />
          <label
            class="m_1738fcb2 mantine-SegmentedControl-label"
            for=""
            style=""
          >
            <span
              class="m_78882f40 mantine-SegmentedControl-innerLabel"
              style=""
            >
              Gelap
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="m_3eebeb36 mantine-Divider-root"
      data-orientation="horizontal"
      role="separator"
      style=""
    />
    <h2
      class="m_8a5d1357 mantine-Title-root"
      data-order="2"
      style="--title-fw: var(--mantine-h2-font-weight); --title-lh: var(--mantine-h2-line-height); --title-fz: var(--mantine-h2-font-size); font-weight: bold;"
    >
      Koleksi Pokemon
    </h2>
    <button
      class="mantine-focus-auto m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
      data-disabled="true"
      disabled=""
      style="--button-bg: var(--mantine-color-red-filled); --button-hover: var(--mantine-color-red-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; max-width: calc(15rem * var(--mantine-scale));"
      type="button"
    >
      <span
        class="m_80f1301b mantine-Button-inner"
        style=""
      >
        <span
          class="m_811560b9 mantine-Button-label"
          style=""
        >
          Hapus Koleksi Pokemon
        </span>
      </span>
    </button>
  </div>
</div>
`;
